File: core/src/galenscovell/oregontrail/processing/pathfinding/Pathfinder.java
package galenscovell.oregontrail.processing.pathfinding;

import galenscovell.oregontrail.map.*;

import java.util.*;

public class Pathfinder {
    private List<Node> openList, closedList;
    private Node startNode, endNode;

    public List<Point> findPath(Tile startTile, Tile endTile, Tile[][] grid) {
        this.openList = new ArrayList<Node>();
        this.closedList = new ArrayList<Node>();
        this.startNode = new Node(startTile);
        this.endNode = new Node(endTile);

        startNode.setCostFromStart(0);
        startNode.setTotalCost(startNode.getCostFromStart() + heuristic(startNode, endNode));
        openList.add(startNode);

        while (!openList.isEmpty()) {
            Node current = getBestNode();

            if (current.getTile() == endNode.getTile()) {
                endNode = current;
                return tracePath();
            }

            openList.remove(current);
            closedList.add(current);

            for (Point point : current.getTile().getNeighbors()) {
                Tile neighborTile = grid[point.y][point.x];

                if (neighborTile != null && (neighborTile == endTile || neighborTile.isEmpty())) {
                    Node neighborNode = new Node(neighborTile);

                    if (!inList(neighborTile, closedList)) {
                        neighborNode.setTotalCost(current.getCostFromStart() + heuristic(neighborNode, endNode));

                        if (!inList(neighborTile, openList)) {
                            neighborNode.setParent(current);
                            openList.add(neighborNode);
                        } else {
                            if (neighborNode.getCostFromStart() < current.getCostFromStart()) {
                                neighborNode.setCostFromStart(neighborNode.getCostFromStart());
                                neighborNode.setParent(neighborNode.getParent());
                            }
                        }
                    }
                }
            }
        }
        return null;
    }

    private boolean inList(Tile nodeTile, List<Node> nodeList) {
        for (Node node : nodeList) {
            if (node.getTile() == nodeTile) {
                return true;
            }
        }
        return false;
    }

    private Node getBestNode() {
        double minCost = Double.POSITIVE_INFINITY;
        Node bestNode = null;

        for (Node node : openList) {
            double totalCost = node.getCostFromStart() + heuristic(node, endNode);
            if (minCost > totalCost) {
                minCost = totalCost;
                bestNode = node;
            }
        }
        return bestNode;
    }

    private double heuristic(Node start, Node end) {
        double dx = start.getTile().x - end.getTile().x;
        double dy = start.getTile().y - end.getTile().y;
        return Math.sqrt(dx * dx + dy * dy);
    }

    private List<Point> tracePath() {
        // Returns ordered list of points along movement path
        List<Point> path = new ArrayList<Point>();
        // Chase parent of node until start point reached
        Node node = endNode;
        while (node.getParent() != null) {
            path.add(new Point(node.getTile().x, node.getTile().y));
            node = node.getParent();
        }
        path.add(new Point(startNode.getTile().x, startNode.getTile().y));
        System.out.println(node.getTotalCost());
        return path;
    }
}

File: core/src/galenscovell/hinterstar/util/Pathfinder.scala
package galenscovell.hinterstar.util

import galenscovell.hinterstar.generation.interior.Tile
import galenscovell.hinterstar.processing.Node

import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer


object Pathfinder {
  private val openList: ArrayBuffer[Node] = ArrayBuffer()
  private val closedList: ArrayBuffer[Node] = ArrayBuffer()



  def findPath(startTile: Tile, endTile: Tile): mutable.Stack[Tile] = {
    openList.clear()
    closedList.clear()

    val startNode: Node = new Node(startTile)
    val endNode: Node = new Node(endTile)

    startNode.setCostFromStart(0)
    startNode.setTotalCost(startNode.getCostFromStart + heuristic(startNode, endNode))
    openList.append(startNode)

    while (openList.nonEmpty) {
      val current: Node = getBestNode(endNode)

      if (current.getTile == endNode.getTile) {
        return tracePath(current)
      }

      openList.remove(openList.indexOf(current))
      closedList.append(current)

      for (neighborTile: Tile <- current.getTile.getNeighbors) {
        if (neighborTile.isTraversible) {
          val neighborNode: Node = new Node(neighborTile)

          if (!inList(neighborTile, closedList)) {
            neighborNode.setTotalCost(current.getCostFromStart + heuristic(neighborNode, endNode))

            if (!inList(neighborTile, openList)) {
              neighborNode.setParent(current)
              openList.append(neighborNode)
            } else {
              if (neighborNode.getCostFromStart < current.getCostFromStart) {
                neighborNode.setCostFromStart(neighborNode.getCostFromStart)
                neighborNode.setParent(neighborNode.getParent)
              }
            }
          }
        }
      }
    }
    null
  }

  private def heuristic(start: Node, end: Node): Double = {
    val xs: Double = (start.getTile.getTx - end.getTile.getTx) * (start.getTile.getTx - end.getTile.getTx)
    val ys: Double = (start.getTile.getTy - end.getTile.getTy) * (start.getTile.getTy - end.getTile.getTy)
    Math.sqrt(xs + ys)
  }

  private def inList(nodeTile: Tile, nodeList: ArrayBuffer[Node]): Boolean = {
    for (n: Node <- nodeList) {
      if (n.getTile == nodeTile) {
        return true
      }
    }

    false
  }

  private def getBestNode(endNode: Node): Node = {
    var minCost: Double = Double.PositiveInfinity
    var bestNode: Node = null

    for (n: Node <- openList) {
      val totalCost: Double = n.getCostFromStart + heuristic(n, endNode)
      if (minCost > totalCost) {
        minCost = totalCost
        bestNode = n
      }
    }

    bestNode
  }

  private def tracePath(n: Node): mutable.Stack[Tile] = {
    val path: mutable.Stack[Tile] = mutable.Stack()
    var node: Node = n

    while (node.getParent != null) {
      path.push(node.getTile)
      node = node.getParent
    }

    path
  }
}
